Update kubernetes to klog v2

dims opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

**Reviews**

sig-auth:
- [x] golang.org/x/oauth2

sig-instrumentation
- [x] github.com/go-logr/logr
- [x] go.opencensus.io
- [x] k8s.io/klog/v2

sig-apimachinery:
- [x] k8s.io/kube-openapi
- [x] k8s.io/gengo

sig-node:
- [x] github.com/google/cadvisor
- [x] github.com/golang/groupcache

sig-cloudprovider:
- [x] cloud.google.com/go
- [x] github.com/googleapis/gax-go
- [x] github.com/GoogleCloudPlatform/k8s-cloud-provider
- [x] google.golang.org/api
- [x] sigs.k8s.io/apiserver-network-proxy

Miscellaneous:
- [x] k8s.io/utils


**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
klog v2 has the new APIs needed for the [structured logging KEP](https://github.com/kubernetes/enhancements/pull/1617)

**Which issue(s) this PR fixes**:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #89961 

**Special notes for your reviewer**:

_New Dependencies_
- github.com/go-logr/logr
- github.com/googleapis/gax-go/v2

**Does this PR introduce a user-facing change?**:
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

dims committed:
Fix for API change in cadvisor

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dashpole commented on pkg/kubelet/stats/stats_provider_test.go:
```diff
@@ -556,8 +556,6 @@ func generateCustomMetricSpec() []cadvisorapiv1.MetricSpec {
 			case 1:
 				e.Type = cadvisorapiv1.MetricCumulative
-			case 2:
-				e.Type = cadvisorapiv1.MetricDelta
```
Hmmm... You removed this in https://github.com/google/cadvisor/pull/2445, but I think it may be possible to use it, since we  cast a string to a metric type here: https://github.com/google/cadvisor/blob/master/collector/prometheus_collector.go#L153

dims replied:
Ah right. please check the new update to see if it is enough. thanks @dashpole

dims commented:
/test pull-kubernetes-e2e-kind

dashpole commented:
cadvisor updated looks good

fedebongio commented:
/cc @logicalhan

dims commented:
@liggitt here are the biggest culprits according to diffstat for lines added:

```
b/vendor/google.golang.org/api/compute/v1/compute-gen.go                              |85711 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------------------------
b/vendor/google.golang.org/api/compute/v1/compute-api.json                            |29009 ++++++++++++++++++++++++--------------
b/vendor/google.golang.org/api/compute/v0.alpha/compute-gen.go                        |25842 ++++++++++++++++++++++++----------
b/vendor/github.com/GoogleCloudPlatform/k8s-cloud-provider/pkg/cloud/gen.go           |16500 +++++++++++++---------
b/vendor/google.golang.org/api/compute/v0.alpha/compute-api.json                      |16152 +++++++++++++--------
b/vendor/google.golang.org/api/compute/v0.beta/compute-gen.go                         |13592 ++++++++++++++++--
b/vendor/google.golang.org/api/compute/v0.beta/compute-api.json                       | 5883 ++++++-
```

full output here : http://paste.openstack.org/raw/792316/

liggitt commented on cluster/images/conformance/BUILD:
```diff
@@ -34,7 +34,7 @@ multi_arch_container(
     docker_tags = ["{{STABLE_DOCKER_REGISTRY}}/conformance-{ARCH}:{{STABLE_DOCKER_TAG}}"],
     env = {
-        "E2E_FOCUS": "\[Conformance\]",
+        "E2E_FOCUS": "\\[Conformance\\]",
```
was this a pre-existing bug?

dims replied:
looks like it.

liggitt commented on go.mod:
```diff
@@ -532,34 +551,35 @@ replace (
-	k8s.io/gengo => k8s.io/gengo v0.0.0-20200205140755-e0e292d8aa12
+	k8s.io/gengo => k8s.io/gengo v0.0.0-20200413195148-3a45101e95ac
 	k8s.io/heapster => k8s.io/heapster v1.2.0-beta.1
 	k8s.io/klog => k8s.io/klog v1.0.0
```
isn't it problematic for us to still have things in our tree that use klog/v1?

liggitt replied:
which version of klog do the CLI flags bind to?

dims replied:
Looks like i may be missing a commit here. let me check

liggitt replied:
klog is pretty pervasive, I expect we'll need a verify check that prevents reintroduction of klog/v1 usage. Will this also become a change all our library consumers have to react to in 1.19 (sweep their own code-bases and all their dependencies to unify on klog/v2)?

dims replied:
so we no longer have klog v1 in our vendor/ with the latest updates i made in this PR

we will have to think through this a bit. folks using us as a library (and currently using klog v1 may have a way to get things to work). I have to try it out and add an example/doc somewhere

liggitt commented:
ok, those line additions are annoying, but not entangling, which is good.

I broke down the new dependencies by source:

things in our dep tree that are not vendored:
```
(via github.com/golangci/golangci-lint, via k8s.io/repo-infra, because we need `kazel`)
    github.com/bombsimon/wsl/v2
    github.com/uudashr/gocognit
    github.com/gofrs/flock
    github.com/jingyugao/rowserrcheck
        github.com/jmoiron/sqlx
            github.com/go-sql-driver/mysql
            github.com/lib/pq
            github.com/mattn/go-sqlite3
    github.com/jirfag/go-printf-func-name
    github.com/securego/gosec
        github.com/lib/pq
    github.com/ultraware/whitespace
    github.com/uudashr/gocognit

(via github.com/google/pprof)
    github.com/chzyer/logex
    github.com/chzyer/readline
    github.com/chzyer/test
    github.com/ianlancetaylor/demangle

(via viper)
    github.com/coreos/bbolt
    github.com/ugorji/go

(via github.com/go-critic/go-critic)
    github.com/pborman/uuid
```

things added to our dep tree we do vendor:
```
(via etcd)
    github.com/mattn/go-colorable
(via viper)
    github.com/subosito/gotenv
    gopkg.in/ini.v1
```

The github.com/golangci/golangci-lint / k8s.io/repo-infra dependency is the most problematic one. I don't think it's reasonable to add all of those dependencies. We only need repo-info for kazel/bazel (if  #88553 was completed we could drop it entirely).

dims committed:
switch over k/k to use klog v2

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims commented:
/test pull-kubernetes-e2e-gce

dims commented:
/test pull-kubernetes-e2e-gce-iscsi

dims commented:
![image](https://user-images.githubusercontent.com/23304/80021771-06676400-84a9-11ea-9a17-dd5a97fcc6e3.png)

https://kubernetes.slack.com/archives/CG3518SFJ/p1587544797016000

dims commented:
/test pull-kubernetes-bazel-test
/test pull-kubernetes-e2e-azure-file-windows
/test pull-kubernetes-e2e-gce

dims commented:
/assign @enj 
(for sig-auth)

/assign @serathius 
(for sig-instrumentation)

/assign @liggitt 
(for sig-apimachinery)

/assign @derekwaynecarr
(for sig-node)

/assign @cheftako  @andrewsykim
(for sig-cloudprovider)

/assign @neolit123 
(for miscellaneous)

neolit123 reviewed (commented):
/approve

dims commented:
/retest

dims commented:
/retest

dims commented:
/test pull-kubernetes-verify
/test pull-kubernetes-integration
/test pull-kubernetes-e2e-kind-ipv6
/test pull-kubernetes-e2e-gci-gce-ipvs
/test pull-kubernetes-e2e-gce-storage-slow

BenTheElder commented:
/retest

serathius commented:
/approve

dims commented:
The bazel test failures are caused by (oauth2: turn Transport.CancelRequest into a no-op)[https://github.com/golang/oauth2/commit/858c2ad4c8b6c5d10852cb89079f6ca1c7309787]:
https://github.com/golang/oauth2/compare/0f29369cfe45...858c2ad4c8b6

Looks like we just have to drop these tests (cc @mikedanese - as Mike added these in https://github.com/kubernetes/kubernetes/pull/71757)

dims committed:
Drop tests that break with newer oauth2

The bazel test failures are caused by (oauth2: turn Transport.CancelRequest into a no-op)[https://github.com/golang/oauth2/commit/858c2ad4c8b6c5d10852cb89079f6ca1c7309787]:
https://github.com/golang/oauth2/compare/0f29369cfe45...858c2ad4c8b6

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

mikedanese commented:
> Looks like we just have to drop these tests (cc @mikedanese - as Mike added these in #71757)

Ack, this is fine. That type of request cancellation is basically deprecated and we have context plumbed now.

dims commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

dims commented:
/test pull-kubernetes-verify 
/test pull-kubernetes-node-e2e

dims commented:
The groupcache change is just one new public method that we don't use:
https://github.com/golang/groupcache/compare/02826c3e7903...215e87163ea7#diff-cb30936e2e31904e87abc8eaeacae028R25-R133

dims commented:
@cheftako looks like we just need sign off from you

```
OLD: google.golang.org/api => google.golang.org/api v0.6.1-0.20190607001116-5213b8090861
NEW: google.golang.org/api => google.golang.org/api v0.15.1-0.20200106000736-b8fc810ca6b5

OLD: google.golang.org/appengine => google.golang.org/appengine v1.5.0
NEW: google.golang.org/appengine => google.golang.org/appengine v1.6.5

OLD: google.golang.org/genproto => google.golang.org/genproto v0.0.0-20191230161307-f3c370f40bfb
NEW: google.golang.org/genproto => google.golang.org/genproto v0.0.0-20200115191322-ca5a22157cba

OLD: github.com/GoogleCloudPlatform/k8s-cloud-provider => github.com/GoogleCloudPlatform/k8s-cloud-provider v0.0.0-20190822182118-27a4ced34534
NEW: github.com/GoogleCloudPlatform/k8s-cloud-provider => github.com/GoogleCloudPlatform/k8s-cloud-provider v0.0.0-20200415212048-7901bc822317

OLD: sigs.k8s.io/apiserver-network-proxy/konnectivity-client => sigs.k8s.io/apiserver-network-proxy/konnectivity-client v0.0.7
NEW: sigs.k8s.io/apiserver-network-proxy/konnectivity-client => sigs.k8s.io/apiserver-network-proxy/konnectivity-client v0.0.9-0.20200506181510-5ab9684eaacf
```

Also if we can get a tag on `sigs.k8s.io/apiserver-network-proxy` that would be great!

dims commented:
/test pull-kubernetes-e2e-kind-ipv6

BenTheElder commented:
/retest

liggitt commented on vendor/github.com/google/cadvisor/container/libcontainer/handler.go:
```diff
@@ -35,7 +36,18 @@ import (
+
+var (
+	whitelistedUlimits      = [...]string{"max_open_files"}
+	referencedResetInterval = flag.Uint64("referenced_reset_interval", 0,
```
this registers a global flag... does this show up in any kube component flags?

liggitt commented on vendor/github.com/google/cadvisor/info/v1/metric.go:
```diff
@@ -27,9 +27,6 @@ const (
 	MetricCumulative MetricType = "cumulative"
-
-	// Rate over a time period.
-	MetricDelta MetricType = "delta"
```
should this be restored in a follow-up, given we're still using it?

caesarxuchao commented:
@dims we will cut a new network proxy version in a few days. I'll update here when it's done.

neolit123 commented:
does go modules allow some replace magic to keep the imports in all files as `k8s.io/klog` (avoiding huge diff), but the underlying import to be `k8s.io/klog/v2`?

liggitt commented:
> does go modules allow some replace magic to keep the imports in all files as `k8s.io/klog` (avoiding huge diff), but the underlying import to be `k8s.io/klog/v2`?

definitely not. the klog API changed in v2

liggitt commented:
fyi - three additions/changes in init funcs:

```
+func init() {
+	nvmLibMutex.Lock()
+	defer nvmLibMutex.Unlock()
+	cErr := C.nvm_init()
+	if cErr != C.NVM_SUCCESS {
+		// Unfortunately klog does not seem to work here. I believe it's better to
+		// output information using fmt rather then let it disappear silently.
+		fmt.Printf("libipmctl initialization failed with status %d", cErr)
+	}
+	isNVMLibInitialized = true
+}

+func init() {
+	libpmfMutex.Lock()
+	defer libpmfMutex.Unlock()
+	pErr := C.pfm_initialize()
+	if pErr != C.PFM_SUCCESS {
+		fmt.Printf("unable to initialize libpfm: %d", int(pErr))
+		return
+	}
+	isLibpfmInitialized = true
+}

@@ -55,13 +56,31 @@ func init() {
 		host = shortHostname(h)
 	}
 
-	current, err := user.Current()
-	if err == nil {
-		userName = current.Username
-	}
+	// On Windows, the Go 'user' package requires netapi32.dll.
+	// This affects Windows Nano Server:
+	//   https://github.com/golang/go/issues/21867
+	// Fallback to using environment variables.
+	if runtime.GOOS == "windows" {
+		u := os.Getenv("USERNAME")
+		if len(u) == 0 {
+			return
+		}
+		// Sanitize the USERNAME since it may contain filepath separators.
+		u = strings.Replace(u, `\`, "_", -1)
 
-	// Sanitize userName since it may contain filepath separators on Windows.
-	userName = strings.Replace(userName, `\`, "_", -1)
+		// user.Current().Username normally produces something like 'USERDOMAIN\USERNAME'
+		d := os.Getenv("USERDOMAIN")
+		if len(d) != 0 {
+			userName = d + "_" + u
+		} else {
+			userName = u
+		}
+	} else {
+		current, err := user.Current()
+		if err == nil {
+			userName = current.Username
+		}
+	}
 }
 
```

liggitt commented:
swept updates, mechanics look reasonable to me

as part of the release note for 1.19, I would expect us to provide guidance for consumers of k8s components:
1. how to update their main() func to switch to klog/v2 setup
2. if they have other dependencies out of their control that are still logging to klog/v1, how to funnel those klog/v1 calls to klog/v2 (sort of what https://github.com/kubernetes/klog/pull/142 is doing, but I would go in the opposite direction and push them to update the code under their control to klog/v2, and set up a klog v1 -> v2 shim to capture log output from legacy dependencies)

dims commented:
Got ack from @cheftako on slack!

dims committed:
Updating dependency k8s.io/gengo to version 8167cfd

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency k8s.io/kube-openapi to version 656914f

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency github.com/GoogleCloudPlatform/k8s-cloud-provider to version 7901bc8

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency k8s.io/utils to version 2df71eb

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency github.com/google/cadvisor to version 6a8d614

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency sigs.k8s.io/apiserver-network-proxy/konnectivity-client to version 33b9978

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency cloud.google.com/go to version v0.51.0

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency google.golang.org/genproto to version v0.0.0-20200115191322-ca5a22157cba

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency github.com/golang/groupcache to version v0.0.0-20191227052852-215e87163ea7

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency github.com/google/pprof to version v0.0.0-20191218002539-d4f498aebedc

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency github.com/jstemmer/go-junit-report to version v0.9.1

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency go.opencensus.io to version v0.22.2

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency google.golang.org/appengine to version v1.6.5

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency honnef.co/go/tools to version v0.0.1-2019.2.3

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency github.com/ianlancetaylor/demangle to version v0.0.0-20181102032728-5e5cf60278f6

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency golang.org/x/exp to version v0.0.0-20191227195350-da58074b4299

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency golang.org/x/lint to version v0.0.0-20191125180803-fdd1cda4f05f

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency golang.org/x/oauth2 to version v0.0.0-20191202225959-858c2ad4c8b6

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency golang.org/x/image to version v0.0.0-20190802002840-cff245a6509b

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency golang.org/x/mobile to version v0.0.0-20190719004257-d2bd2a29d028

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Updating dependency golang.org/x/mod to version v0.1.0

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

dims committed:
Run hack/update-vendor.sh

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

cheftako commented:
/lgtm with @liggitt suggested documentation

cheftako commented:
/retest

dims committed:
fix API change in apiserver-network-proxy

Signed-off-by: Davanum Srinivas <davanum@gmail.com>

liggitt commented:
/lgtm
/approve

dims commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-kubemark-e2e-gce-big

dims commented:
/retest

dims commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

dims commented:
/retest

dims commented:
/test pull-kubernetes-integration

dims commented:
/retest

dims commented:
/retest

dims commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

caesarxuchao commented:
@dims we created v0.0.9 tag for the network proxy repo: https://github.com/kubernetes-sigs/apiserver-network-proxy/tree/v0.0.9

dims commented:
/test pull-kubernetes-e2e-gci-gce-ipvs

BenTheElder commented:
bah, "GCI" hasn't been meaningful for a long time. :/

dims commented:
/test pull-kubernetes-e2e-gce

dims commented:
@neolit123 @liggitt can one of you please reapply lgtm?

liggitt commented:
/lgtm

dims commented:
good morning @liggitt !!

dims commented:
/test pull-kubernetes-e2e-gce 
/test pull-kubernetes-verify

liggitt commented:
/skip

dims commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

liggitt commented:
Hmm

> The pod could not start because it could not mount the volume "gcs-credentials": couldn't propagate object cache: timed out waiting for the condition

liggitt commented:
/retest

dims commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-integration

liggitt commented:
> The pod could not start because it could not mount the volume "gcs-credentials": couldn't propagate object cache: timed out waiting for the condition

Saw this again. @BenTheElder, this ring a bell?

dims commented:
/test pull-kubernetes-e2e-kind
/test pull-kubernetes-integration

krmayankk commented:
@dims  so which of the kubernetes components moved to json or is this just moving from klog to klog v2 and the json apis will be used selectively later ?
